### PR TITLE
transformer int4 and native int4's benchmark script for 32 256 1k 2k input

### PR DESCRIPTION
**Pull request by qiuxin2012**
## Description

transformer int4 and native int4's benchmark script for 32 256 1k 2k input

### 1. Why the change?

transformer int4 and native int4's benchmark script for 32 256 1k 2k input

### 2. User API changes

None

### 3. Summary of the change 

Usage:
```
python native_benchmark.py --repo-id meta-llama/Llama-2-7b-chat-hf --local-model-hub ~/IntelAnalytics/models/ --prompt-len 2048  --infer-times 2  --save-path /tmp/llama-7b

python transformers_benchmark.py --repo-id 'THUDM/chatglm2-6b' --local-model-hub ~/IntelAnalytics/models --prompt-len 1024 --max-tokens 128   --infer-times 2
```
The scripts will load models from local-model-hub if it's provided.

### 4. How to test?
- [ ] N/A
- [ ] Unit test
- [ ] Application test
- [ ] Document test
- [ ] ...

### 5. New dependencies

- [ ] New Python dependencies
       - Dependency1 
       - Dependency2
       - ...
- [ ] New Java/Scala dependencies and their license
       - Dependency1 and license1
       - Dependency2 and license2
       - ...

### CODE REVIEW

**Review by Litchilitchy — APPROVED** [review]

**Comment by jason-dai** [comment]
We need input length upto 8K

File: python/llm/dev/benchmark/pipelines/transformers_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    end = time.time()
+    print(">> loading and conversion of model costs {}s".format(end-st))
+
+    model = BenchmarkWrapper(model)
+
+    input_str = open(f"prompt/{args.prompt_len}.txt", 'r').read()
+
+    with torch.inference_mode():
+        for i in range(infer_times):
+            input_ids = tokenizer.encode(input_str, return_tensors="pt")
```
Comment (hkvision) [inline python/llm/dev/benchmark/pipelines/transformers_benchmark.py:74]: For encode we only need to run it once?

File: python/llm/dev/benchmark/pipelines/transformers_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    st = time.time()
+    # TODO: auto detect using config.json
+    if repo_id in ['THUDM/chatglm-6b', 'THUDM/chatglm2-6b']:
+        model = AutoModel.from_pretrained(model_path, load_in_4bit=True, trust_remote_code=True, torch_dtype='auto')
+        tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
+    else:
+        model = AutoModelForCausalLM.from_pretrained(model_path, load_in_4bit=True)
+        end = time.time()
+        print(">> loading of model costs {}s".format(end-st))
+        tokenizer = LlamaTokenizer.from_pretrained(model_path)
```
Comment (hkvision) [inline python/llm/dev/benchmark/pipelines/transformers_benchmark.py:63]: Why don't we use LlamaTokenizer here?

File: python/llm/dev/benchmark/pipelines/transformers_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Transformer INT4 example')
+    parser.add_argument('--repo-id', type=str, required=True,
+                        help='The huggingface repo id for the large language model to be downloaded'
+                             ', or the path to the huggingface checkpoint folder')
+    parser.add_argument('--local-model-hub', type=str,
+                        help='the local caching hub dir.')
+    parser.add_argument('--prompt-len', type=int, default=32,
+                        help='prompt length to infer')
+    parser.add_argument('--infer-times', type=int, default=3,
```
Comment (hkvision) [inline python/llm/dev/benchmark/pipelines/transformers_benchmark.py:39]: rename to num-trials?

File: python/llm/dev/benchmark/pipelines/transformers_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    with torch.inference_mode():
+        for i in range(infer_times):
+            input_ids = tokenizer.encode(input_str, return_tensors="pt")
+            # As different tokenizer has different encodings,
+            # slice the input_ids to ensure the prompt length is required length.
+            input_ids = input_ids[:, :args.prompt_len]
+            print("input length is : ", input_ids.shape[1])
+            st = time.perf_counter()
+            output_ids = model.generate(input_ids, do_sample=False, max_new_tokens=max_tokens)
```
Comment (hkvision) [inline python/llm/dev/benchmark/pipelines/transformers_benchmark.py:80]: Is there any way to make sure the max_tokens are reached? e.g. check len(output_ids), if not reached, we need to re-run this trial

File: python/llm/dev/benchmark/pipelines/transformers_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            end = time.perf_counter()
+            print("model generate cost: " + str(end - st))
```
Comment (hkvision) [inline python/llm/dev/benchmark/pipelines/transformers_benchmark.py:82]: shall we print this out? Better not to make it public

Comment (Litchilitchy) [reply python/llm/dev/benchmark/pipelines/transformers_benchmark.py:82]: There's a param controlling this. If not public scenario, add `do_print=False` in `BenchmarkWrapper(do_print=False)`

File: python/llm/dev/benchmark/pipelines/transformers_benchmark.py (hunk 4 above)
Comment (jason-dai) [reply python/llm/dev/benchmark/pipelines/transformers_benchmark.py:80]: > Is there any way to make sure the max_tokens are reached? e.g. check len(output_ids), if not reached, we need to re-run this trial

No, I don't think we can enforce that

Comment (hkvision) [reply python/llm/dev/benchmark/pipelines/transformers_benchmark.py:80]: okay

File: python/llm/dev/benchmark/pipelines/transformers_benchmark.py (hunk 5 above)
Comment (hkvision) [reply python/llm/dev/benchmark/pipelines/transformers_benchmark.py:82]: But is this parameter configurable in this script?

**Comment by qiuxin2012** [comment]
> We need input length upto 8K

Will do it later.

**Pull request merged**